File: study/lua/src/app/scene/view/guild/GuildHallBaseInfoNode.lua
--
-- Author: Liangxu
-- Date: 2017-06-22 14:10:42
-- 军团大厅基本信息
local ViewBase = require("app.ui.ViewBase")
local GuildHallBaseInfoNode = class("GuildHallBaseInfoNode", ViewBase)
local UserDataHelper = require("app.utils.UserDataHelper")
local PopupGuildAnnouncement = require("app.scene.view.guild.PopupGuildAnnouncement")
local GuildConst = require("app.const.GuildConst")
local PopupSystemAlert = require("app.ui.PopupSystemAlert")
local ParameterIDConst = require("app.const.ParameterIDConst")
local TextHelper = require("app.utils.TextHelper")
local GuildTaskItemCell = require("app.scene.view.guild.GuildTaskItemCell")
local GuildTaskBoxNodeHelper = require("app.scene.view.guild.GuildTaskBoxNodeHelper")
local LogicCheckHelper  = require("app.utils.LogicCheckHelper")
local DataConst = require("app.const.DataConst")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")

function GuildHallBaseInfoNode:ctor()
	self._listItemSource2 = nil
	self._taskBoxHelper = nil
	self._imageNameBg = nil
	local resource = {
		file = Path.getCSB("GuildHallBaseInfoNode", "guild"),
		binding = {
			_buttonLogout = {
				events = {{event = "touch", method = "_onButtonLogout"}}
			},
			_buttonModifyAnnouncement = {
				events = {{event = "touch", method = "_onButtonModifyAnnouncement"}}
			},
			_buttonModifyDeclaration = {
				events = {{event = "touch", method = "_onButtonModifyDeclaration"}}
			},
			_imageNameBg = {
				events = {{event = "touch", method = "_onButtonModifyName"}}	
			},
		}
	}
	GuildHallBaseInfoNode.super.ctor(self, resource)
end

function GuildHallBaseInfoNode:onCreate()
	--self._buttonModifyAnnouncement:setString(Lang.get("guild_btn_modify"))
	--self._buttonModifyDeclaration:setString(Lang.get("guild_btn_modify"))
	self._listItemSource2:setTemplate(GuildTaskItemCell)
	self._listItemSource2:setCallback(handler(self, self._onItemUpdate), handler(self, self._onItemSelected))
	self._listItemSource2:setCustomCallback(handler(self, self._onItemTouch))

	self._taskBoxHelper = GuildTaskBoxNodeHelper.new(self._boxNode)
end

function GuildHallBaseInfoNode:onEnter()
	self._signalGuildBaseInfoUpdate = G_SignalManager:add(SignalConst.EVENT_GUILD_BASE_INFO_UPDATE, handler(self, self._onEventGuildBaseInfoUpdate))
	self._signalSetGuildMessage = G_SignalManager:add(SignalConst.EVENT_GUILD_SET_MESSAGE_SUCCESS, handler(self, self._onEventSetGuildMessage))
	self._signalGuildBoxReward = G_SignalManager:add(SignalConst.EVENT_GUILD_BOX_REWARD, handler(self, self._onEventGuildBoxReward))
	self._signalGuildNameChange = G_SignalManager:add(SignalConst.EVENT_GUILD_NAME_CHANGE, handler(self, self._onEventGuildNameChange))
	self._signalGuildGetUserGuild = G_SignalManager:add(SignalConst.EVENT_GUILD_GET_USER_GUILD, handler(self, self._onEventGuildGetUserGuild))
end

function GuildHallBaseInfoNode:onExit()
	self._signalGuildBaseInfoUpdate:remove()
	self._signalGuildBaseInfoUpdate = nil
	self._signalSetGuildMessage:remove()
	self._signalSetGuildMessage = nil

	self._signalGuildBoxReward:remove()
	self._signalGuildBoxReward = nil

	self._signalGuildNameChange:remove()
	self._signalGuildNameChange = nil
	
	self._signalGuildGetUserGuild:remove()
	self._signalGuildGetUserGuild = nil
end

function GuildHallBaseInfoNode:updateView()
	G_UserData:getGuild():c2sGetGuildBase()	
end

function GuildHallBaseInfoNode:_updateInfo()
	self._myGuild = G_UserData:getGuild():getMyGuild()
	assert(self._myGuild, "G_UserData:getGuild():getMyGuild() = nil")

	local icon = self._myGuild:getIcon()
	local iconRes = Path.getCommonIcon("guild", icon)
	local name = self._myGuild:getName()
	local level = G_UserData:getGuild():getMyGuildLevel()--self._myGuild:getLevel()
	local exp = G_UserData:getGuild():getMyGuildExp()--self._myGuild:getExp()
	local needExp = UserDataHelper.getGuildLevelUpNeedExp(level)
	local names = UserDataHelper.getGuildLeaderNames()
	local leaderName = names.leaderName or Lang.get("guild_leader_name_default")
	local mateName = names.mateName or Lang.get("guild_leader_name_default")
	local elderName = names.elderNames and names.elderNames[1] or Lang.get("guild_leader_name_default")
	local elderName2 = names.elderNames and names.elderNames[2] or Lang.get("guild_leader_name_default")
	local memberNumber = self._myGuild:getMember_num()
	local maxMember = UserDataHelper.getGuildMaxMember(level)
	local announcement = UserDataHelper.getGuildAnnouncement(self._myGuild)  
	local declaration = UserDataHelper.getGuildDeclaration(self._myGuild) 
	local guildRank = self._myGuild:getGuild_rank()

	--self._imageIcon:loadTexture(iconRes)
	self._textName:setString(name)
	self._textLevel:setString(Lang.get("guild_hall_level", {level = level}))
	self._loadingBarProgress:setPercent(exp / needExp * 100)
	self._textProgress:setString(exp.."/"..needExp)
	self._textLeaderName:setString(leaderName)
	self._textMateName:setString(mateName)
	self._textElderName:setString(elderName)
	self._textElderName2:setString(elderName2)
	self._textMemberNumber:setString(memberNumber.."/"..maxMember)
	self._textAnnouncement:setString(announcement)
	self._textDeclaration:setString(declaration)
	self._textRank:setString(tostring(guildRank))

	self:_updateBtnState()
	self:_updateList()
	self._taskBoxHelper:refreshBoxView()
end

function GuildHallBaseInfoNode:_updateBtnState()
	local userMemberData = G_UserData:getGuild():getMyMemberData()
	local myPosition = userMemberData:getPosition()
	self._canDissolve = UserDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_1) --是否有解散军团的权限
	self._canSetAnnouncement = UserDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_7) --是否能修改公告
	self._canSetDeclaration = UserDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_8) --是否能修改宣言
	self._canModifyGuildName = UserDataHelper.isHaveJurisdiction(myPosition,GuildConst.GUILD_JURISDICTION_10) --是否能修改军团名

	if self._canDissolve then
		self._buttonLogout:setString(Lang.get("guild_btn_dissolve"))
	else
		self._buttonLogout:setString(Lang.get("guild_btn_logout"))
	end
	self._buttonModifyAnnouncement:setVisible(self._canSetAnnouncement)
	self._buttonModifyDeclaration:setVisible(self._canSetDeclaration)

	self._editName:setVisible(self._canModifyGuildName)
	self._imageNameBg:setTouchEnabled(self._canModifyGuildName)
end

--退出\解散
function GuildHallBaseInfoNode:_onButtonLogout()
	if self._canDissolve then
		local count = G_UserData:getGuild():getGuildMemberCount()
		if count > 1 then
			G_Prompt:showTip(Lang.get("guild_tip_can_not_dissolve"))
			return
		end

		local function callbackOK()
			G_UserData:getGuild():c2sGuildDismiss()
		end
		local content = Lang.get("guild_dissolve_hint")
		local title = Lang.get("guild_appoint_confirm_title")
		local popup = require("app.ui.PopupAlert").new(title, content, callbackOK)
		popup:openWithAction()
	else
		local userMemberData = G_UserData:getGuild():getMyMemberData()
		local time = userMemberData:getTime()
		if not UserDataHelper.checkCanQuitGuild(time) then
			return
		end

		local timeLimit = tonumber(require("app.config.parameter").get(ParameterIDConst.GUILD_QUIT_CD_ID).content)
		local timeStr = G_ServerTime:getDayOrHourOrMinFormat(timeLimit)
		local content = Lang.get("guild_leave_hint", {time = timeStr})
		local title = Lang.get("guild_appoint_confirm_title")
		local function callbackOK()
			G_UserData:getGuild():c2sGuildLeave()
		end

		local popup = require("app.ui.PopupAlert").new(title, content, callbackOK)
		popup:openWithAction()
	end
end

--修改公告
function GuildHallBaseInfoNode:_onButtonModifyAnnouncement()
	local popup = PopupGuildAnnouncement.new(handler(self, self._onSaveAnnouncement))
	popup:setTitle(Lang.get("guild_title_announcement"))
	local content =  UserDataHelper.getGuildAnnouncement( self._myGuild) 
	popup:setContent(content)
	popup:openWithAction()
end

--修改宣言
function GuildHallBaseInfoNode:_onButtonModifyDeclaration()
	local popup = PopupGuildAnnouncement.new(handler(self, self._onSaveDeclaration))
	popup:setTitle(Lang.get("guild_title_declaration"))
	local content = UserDataHelper.getGuildDeclaration(self._myGuild)  
	popup:setContent(content)
	popup:openWithAction()
end

--保存公告
function GuildHallBaseInfoNode:_onSaveAnnouncement(content)
	G_UserData:getGuild():c2sSetGuildMessage(content, GuildConst.GUILD_MESSAGE_TYPE_1)
end

--保存宣言
function GuildHallBaseInfoNode:_onSaveDeclaration(content)
	G_UserData:getGuild():c2sSetGuildMessage(content, GuildConst.GUILD_MESSAGE_TYPE_2)
end

function GuildHallBaseInfoNode:_onEventSetGuildMessage(eventName, type)
	if not G_UserData:getGuild():isInGuild() then
		return
	end
	if type == GuildConst.GUILD_MESSAGE_TYPE_1 then
		local content = UserDataHelper.getGuildAnnouncement() 
		self._textAnnouncement:setString(content)
		G_Prompt:showTip(Lang.get("guild_announcement_change_success"))
	elseif type == GuildConst.GUILD_MESSAGE_TYPE_2 then
		local content =  UserDataHelper.getGuildDeclaration() 
		self._textDeclaration:setString(content)
		G_Prompt:showTip(Lang.get("guild_declaration_change_success"))
	end
end

--改军团名
function GuildHallBaseInfoNode:_onButtonModifyName(sender)
	local callback = function(guildName)
		G_UserData:getGuild():c2sGuildChangeName(guildName)
	end
	local ParameterIDConst = require("app.const.ParameterIDConst")
	local PopupGuildChangeName = require("app.scene.view.guild.PopupGuildChangeName")
	local popup = PopupGuildChangeName.new()
	popup:updateUI(
		Lang.get("guild_change_name_title"),
		Lang.get("guild_change_name_hint"),
		{ type = TypeConvertHelper.TYPE_RESOURCE,value =  DataConst.RES_DIAMOND, 
		size = UserDataHelper.getParameter(ParameterIDConst.GUILD_RENAME_COST)},
		callback)
	popup:openWithAction()
	
end

--解散成功
function GuildHallBaseInfoNode:_dismissSuccess()
	G_Prompt:showTip(Lang.get("guild_tip_dismiss_success"))
	G_SceneManager:popScene()
end


function GuildHallBaseInfoNode:_updateList()
	self._guildTaskList = G_UserData:getGuild():getMyGuild():getSortedTaskDataList()
	self._listItemSource2:clearAll()
	self._listItemSource2:resize(math.ceil(#self._guildTaskList))-- 一行3个
end

function GuildHallBaseInfoNode:_onItemUpdate(item, index)
	if self._guildTaskList[index + 1] then
		item:update(self._guildTaskList[index + 1])
	end
end

function GuildHallBaseInfoNode:_onItemSelected(item, index)
end

function GuildHallBaseInfoNode:_onItemTouch(lineIndex,index)
	local data = self._guildTaskList[index + 1]
	if not data then
		return
	end
	
end

function GuildHallBaseInfoNode:_onEventGuildBoxReward(event,rewards)
	if not G_UserData:getGuild():isInGuild() then
		return
	end
	--刷新宝箱
	self._taskBoxHelper:refreshBoxView()
	if rewards then
		G_Prompt:showAwards(rewards)
	end
end

function GuildHallBaseInfoNode:_onEventGuildBaseInfoUpdate(event)
	if not G_UserData:getGuild():isInGuild() then
		return
	end
	self:_updateInfo()
end

function GuildHallBaseInfoNode:_onEventGuildNameChange(event)
	if not G_UserData:getGuild():isInGuild() then
		return
	end
	G_Prompt:showTip(Lang.get("guild_tip_change_name_success"))
end

function GuildHallBaseInfoNode:_onEventGuildGetUserGuild(event)
	if not G_UserData:getGuild():isInGuild() then
		return
	end
	self:_updateInfo()
end



return GuildHallBaseInfoNode

File: study/ts/script/scene/view/guild/GuildHallBaseInfoNode.ts
const { ccclass, property } = cc._decorator;

import { ConfigNameConst } from '../../../const/ConfigNameConst';
import { DataConst } from '../../../const/DataConst';
import { GuildConst } from '../../../const/GuildConst';
import ParameterIDConst from '../../../const/ParameterIDConst';
import { SignalConst } from '../../../const/SignalConst';
import { G_ConfigLoader, G_Prompt, G_SceneManager, G_ServerTime, G_SignalManager, G_UserData } from '../../../init';
import { Lang } from '../../../lang/Lang';
import CommonButtonEdit from '../../../ui/component/CommonButtonEdit';
import CommonButtonTriggle from '../../../ui/component/CommonButtonTriggle';
import PopupAlert from '../../../ui/PopupAlert';
import { GuildDataHelper } from '../../../utils/data/GuildDataHelper';
import { UserDataHelper } from '../../../utils/data/UserDataHelper';
import { handler } from '../../../utils/handler';
import { Path } from '../../../utils/Path';
import { TypeConvertHelper } from '../../../utils/TypeConvertHelper';
import { Util } from '../../../utils/Util';
import ViewBase from '../../ViewBase';
import GuildTaskBoxNodeHelper from './GuildTaskBoxNodeHelper';
import GuildTaskItemCell from './GuildTaskItemCell';
import PopupGuildAnnouncement from './PopupGuildAnnouncement';
import PopupGuildChangeName from './PopupGuildChangeName';
import CommonButtonLevel0Highlight from '../../../ui/component/CommonButtonLevel0Highlight';
import PopupGuildSetWechatQQ from './PopupGuildSetWechatQQ';


@ccclass
export default class GuildHallBaseInfoNode extends ViewBase {

    @property({
        type: cc.Label,
        visible: true
    })
    _textAnnouncement: cc.Label = null;

    @property({
        type: CommonButtonEdit,
        visible: true
    })
    _buttonModifyAnnouncement: CommonButtonEdit = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textDeclaration: cc.Label = null;

    @property({
        type: CommonButtonEdit,
        visible: true
    })
    _buttonModifyDeclaration: CommonButtonEdit = null;

    @property({
        type: cc.ScrollView,
        visible: true
    })
    _listItemSource2: cc.ScrollView = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _boxNode: cc.Node = null;

    @property({
        type: cc.ProgressBar,
        visible: true
    })
    _loadingBarProgress: cc.ProgressBar = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textProgress: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageNameBg: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textLeaderName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textLevel: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textMemberNumber: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textRank: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _editName: cc.Sprite = null;

    @property({
        type: CommonButtonTriggle,
        visible: true
    })
    _buttonLogout: CommonButtonTriggle = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textElderName2: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textElderName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textMateName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textWxQQ: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textWxQQName: cc.Label = null;

    @property({
        type: CommonButtonLevel0Highlight,
        visible: true
    })
    _btnSet: CommonButtonLevel0Highlight = null;

    _taskBoxHelper: GuildTaskBoxNodeHelper;
    _signalSetGuildMessage: any;
    _signalGuildBaseInfoUpdate: any;
    _signalGuildBoxReward: any;
    _signalGuildNameChange: any;
    _signalGuildGetUserGuild: any;
    _myGuild: any;
    _canDissolve: any;
    _canSetAnnouncement: any;
    _canSetDeclaration: any;
    _canModifyGuildName: any;
    _guildTaskList: any;
    private _signalSetWeChatQQ:any;

    onCreate() {
        // this._listItemSource2.setTemplate(GuildTaskItemCell);
        // this._listItemSource2.setCallback(handler(this, this._onItemUpdate), handler(this, this._onItemSelected));
        // this._listItemSource2.setCustomCallback(handler(this, this._onItemTouch));
        this._taskBoxHelper = new GuildTaskBoxNodeHelper(this._boxNode);
    }
    onEnter() {
        this._signalGuildBaseInfoUpdate = G_SignalManager.add(SignalConst.EVENT_GUILD_BASE_INFO_UPDATE, handler(this, this._onEventGuildBaseInfoUpdate));
        this._signalSetGuildMessage = G_SignalManager.add(SignalConst.EVENT_GUILD_SET_MESSAGE_SUCCESS, handler(this, this._onEventSetGuildMessage));
        this._signalGuildBoxReward = G_SignalManager.add(SignalConst.EVENT_GUILD_BOX_REWARD, handler(this, this._onEventGuildBoxReward));
        this._signalGuildNameChange = G_SignalManager.add(SignalConst.EVENT_GUILD_NAME_CHANGE, handler(this, this._onEventGuildNameChange));
        this._signalGuildGetUserGuild = G_SignalManager.add(SignalConst.EVENT_GUILD_GET_USER_GUILD, handler(this, this._onEventGuildGetUserGuild));
        this._signalSetWeChatQQ = G_SignalManager.add(SignalConst.EVENT_GUILD_SET_WE_QQ_ADDRESS,this._onEventSetAddress.bind(this))
        this._btnSet.addClickEventListenerEx(handler(this,this.onBtnShowSetWxQQ));

        var data =  G_UserData.getGuild().getMyMemberData();
         var pos = data.getPosition();
         var config = G_ConfigLoader.getConfig(ConfigNameConst.GUILD_PURVIEW).get(pos);
         var purview:string = config["purview"]
         var purviewIds = purview.split("|")
         if(purviewIds.indexOf("17")>=0)
         {
            this._isTuanZhang = true;
         }
         else
         {
            this._isTuanZhang = false;
         }

        this.updateWXQQUI();
    }
    private _isTuanZhang:boolean = false;
    private _isSetedWx:boolean = false;
    private _isSetedQQ:boolean = false;
    private updateWXQQUI():void{

        var adress:string = G_UserData.getGuild().getMyGuild().getAddress();
        var adress_type = G_UserData.getGuild().getMyGuild().getAddress_type();

        var str = "";
        if(adress.length>8)
        {
            str = adress.substring(0,7)+"..."
        }
        else
        {
            str = adress;
        }

        if(this._isTuanZhang)
        {
            this._btnSet.setString("设置");
        }
        else
        {
            this._btnSet.setString("复制");
        }
        if(adress=="")
        {
            this._textWxQQName.string = "团长微信:";
            this._textWxQQ.string = "暂未设置";
            this._isSetedWx = false;
            this._isSetedQQ = false;
            
        }
        else if(adress_type==0)
        {
            this._textWxQQName.string = "团长微信:";

            this._textWxQQ.string = str;
            this._isSetedWx = true;
            this._isSetedQQ = false;
            
        }
        else
        {
            this._textWxQQName.string = "军团Q群:";
            this._textWxQQ.string = str;
            this._isSetedWx = false;
            this._isSetedQQ = true;
        }
    }
    onExit() {
        this._signalGuildBaseInfoUpdate.remove();
        this._signalGuildBaseInfoUpdate = null;
        this._signalSetGuildMessage.remove();
        this._signalSetGuildMessage = null;
        this._signalGuildBoxReward.remove();
        this._signalGuildBoxReward = null;
        this._signalGuildNameChange.remove();
        this._signalGuildNameChange = null;
        this._signalGuildGetUserGuild.remove();
        this._signalGuildGetUserGuild = null;
        this._signalSetWeChatQQ.remove();
        this._signalSetWeChatQQ = null;
    }
    private _onEventSetAddress():void{
        this.updateWXQQUI();
    }
    copyContentToClipboard(content:string,tips:string):void{
        if(cc.sys.platform == cc.sys.WECHAT_GAME)
        {
                wx.setClipboardData({
                data: content,
                success: function(res) {
                  // self.setData({copyTip:true}),
                  G_Prompt.showTip(tips);
                }
              });
            
        }
        else 
		{
            var webCopyString = function(str){
                console.log('复制');
            
                var input = str + '';
                const el = document.createElement('textarea');
                el.value = input;
                el.setAttribute('readonly', '');
                el.style.contain = 'strict';
                el.style.position = 'absolute';
                el.style.left = '-9999px';
                el.style.fontSize = '12pt'; // Prevent zooming on iOS
            
                const selection = getSelection();
                var originalRange = false;
                if (selection.rangeCount > 0) {
                    originalRange = selection.getRangeAt(0);
                }
                document.body.appendChild(el);
                el.select();
                el.selectionStart = 0;
                el.selectionEnd = input.length;
            
                var success = false;
                try {
                    success = document.execCommand('copy');
                } catch (err) {}
            
                document.body.removeChild(el);
            
                if (originalRange) {
                    selection.removeAllRanges();
                    selection.addRange(originalRange);
                }
            
                return success;
            }
            if(webCopyString(content))
            {
                G_Prompt.showTip(tips); 
            }
        }

    }
    onBtnShowSetWxQQ():void{
        if(this._isTuanZhang)
        {
            G_SceneManager.showDialog("prefab/guild/PopupGuildSetWechatQQ",function(pop:PopupGuildSetWechatQQ){
               pop.open();
         })
        }
        else
        {
            var adress:string = G_UserData.getGuild().getMyGuild().getAddress();
            if(this._isSetedWx)
            {
                this.copyContentToClipboard(adress,"复制成功,快去添加团长微信叭");
            }
            else if(this._isSetedQQ)
            {
                this.copyContentToClipboard(adress,"复制成功,快去添加军团Q群叭");
            }
            else
            {
                
                G_Prompt.showTip("复制失败，快联系团长让他设置团长微信或军团Q群叭");
            }
        }
    }
    updateView() {
        G_UserData.getGuild().c2sGetGuildBase();
    }
    _updateInfo() {
        this._myGuild = G_UserData.getGuild().getMyGuild();
        // assert(this._myGuild, 'G_UserData:getGuild():getMyGuild() = nil');
        var icon = this._myGuild.getIcon();
        var iconRes = Path.getCommonIcon('guild', icon);
        var name = this._myGuild.getName();
        var level = G_UserData.getGuild().getMyGuildLevel();
        var exp = G_UserData.getGuild().getMyGuildExp();
        var needExp = GuildDataHelper.getGuildLevelUpNeedExp(level);
        var names = GuildDataHelper.getGuildLeaderNames();
        var leaderName = names.leaderName || Lang.get('guild_leader_name_default');
        var mateName = names.mateName || Lang.get('guild_leader_name_default');
        var elderName = names.elderNames && names.elderNames[1] || Lang.get('guild_leader_name_default');
        var elderName2 = names.elderNames && names.elderNames[2] || Lang.get('guild_leader_name_default');
        var memberNumber = this._myGuild.getMember_num();
        var maxMember = GuildDataHelper.getGuildMaxMember(level);
        var announcement = GuildDataHelper.getGuildAnnouncement();
        var declaration = GuildDataHelper.getGuildDeclaration(this._myGuild);
        var guildRank = this._myGuild.getGuild_rank();
        this._textName.string = (name);
        this._textLevel.string = (Lang.get('guild_hall_level', { level: level }));
        this._loadingBarProgress.progress = (exp / needExp);
        this._textProgress.string = (exp + ('/' + needExp));
        this._textLeaderName.string = (leaderName);
        this._textMateName.string = (mateName);
        this._textElderName.string = (elderName);
        this._textElderName2.string = (elderName2);
        this._textMemberNumber.string = (memberNumber + ('/' + maxMember));
        this._textAnnouncement.string = (announcement);
        this._textDeclaration.string = (declaration);
        this._textRank.string = (guildRank.toString());
        this._updateBtnState();
        this._updateList();
        this._taskBoxHelper.refreshBoxView();
    }

    update() {
        let a = this._loadingBarProgress.progress;
    }

    _updateBtnState() {
        var userMemberData = G_UserData.getGuild().getMyMemberData();
        var myPosition = userMemberData.getPosition();
        this._canDissolve = GuildDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_1);
        this._canSetAnnouncement = GuildDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_7);
        this._canSetDeclaration = GuildDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_8);
        this._canModifyGuildName = GuildDataHelper.isHaveJurisdiction(myPosition, GuildConst.GUILD_JURISDICTION_10);
        if (this._canDissolve) {
            this._buttonLogout.setString(Lang.get('guild_btn_dissolve'));
        } else {
            this._buttonLogout.setString(Lang.get('guild_btn_logout'));
        }
        this._buttonModifyAnnouncement.node.active = (this._canSetAnnouncement);
        this._buttonModifyDeclaration.node.active = (this._canSetDeclaration);
        this._editName.node.active = (this._canModifyGuildName);
        // this._imageNameBg.setTouchEnabled(this._canModifyGuildName);
    }

    private callbackOK() {
        G_UserData.getGuild().c2sGuildDismiss();
    }

    private callbackOK2() {
        G_UserData.getGuild().c2sGuildLeave();
    }

    _onButtonLogout() {
        if (this._canDissolve) {
            var count = G_UserData.getGuild().getGuildMemberCount();
            if (count > 1) {
                G_Prompt.showTip(Lang.get('guild_tip_can_not_dissolve'));
                return;
            }
            var content = Lang.get('guild_dissolve_hint');
            var title = Lang.get('guild_appoint_confirm_title');

            var resource = cc.resources.get("prefab/common/PopupAlert");
            var node1 = cc.instantiate(resource) as cc.Node;
            let cell = node1.getComponent(PopupAlert);
            cell.init(title, content, this.callbackOK);
            cell.openWithAction();
        } else {
            var userMemberData = G_UserData.getGuild().getMyMemberData();
            var time = userMemberData.getTime();
            if (!GuildDataHelper.checkCanQuitGuild(time)) {
                return;
            }
            var timeLimit = parseInt(G_ConfigLoader.getConfig(ConfigNameConst.PARAMETER).get(ParameterIDConst.GUILD_QUIT_CD_ID).content);
            var timeStr = G_ServerTime.getDayOrHourOrMinFormat(timeLimit);
            var content = Lang.get('guild_leave_hint', { time: timeStr });
            var title = Lang.get('guild_appoint_confirm_title');

            var resource = cc.resources.get("prefab/common/PopupAlert");
            var node1 = cc.instantiate(resource) as cc.Node;
            let cell = node1.getComponent(PopupAlert);
            cell.init(title, content, this.callbackOK2);
            cell.openWithAction();
        }
    }
    onButtonModifyAnnouncement() {
        var popup: PopupGuildAnnouncement = Util.getNode('prefab/guild/PopupGuildAnnouncement', PopupGuildAnnouncement);
        popup.initData(handler(this, this._onSaveAnnouncement));
        popup.setTitle(Lang.get('guild_title_announcement'));
        popup.openWithAction();
        var content = UserDataHelper.getGuildAnnouncement();
        popup.setContent(content);
    }
    onButtonModifyDeclaration() {
        var popup: PopupGuildAnnouncement = Util.getNode('prefab/guild/PopupGuildAnnouncement', PopupGuildAnnouncement);
        popup.initData(handler(this, this._onSaveDeclaration));
        popup.setTitle(Lang.get('guild_title_declaration'));
        var content = UserDataHelper.getGuildDeclaration();
        popup.setContent(content);
        popup.openWithAction();
    }
    _onSaveAnnouncement(content) {
        G_UserData.getGuild().c2sSetGuildMessage(content, GuildConst.GUILD_MESSAGE_TYPE_1);
    }
    _onSaveDeclaration(content) {
        G_UserData.getGuild().c2sSetGuildMessage(content, GuildConst.GUILD_MESSAGE_TYPE_2);
    }
    _onEventSetGuildMessage(eventName, type) {
        if (!G_UserData.getGuild().isInGuild()) {
            return;
        }
        if (type == GuildConst.GUILD_MESSAGE_TYPE_1) {
            var content = GuildDataHelper.getGuildAnnouncement();
            this._textAnnouncement.string = (content);
            G_Prompt.showTip(Lang.get('guild_announcement_change_success'));
        } else if (type == GuildConst.GUILD_MESSAGE_TYPE_2) {
            var content = GuildDataHelper.getGuildDeclaration(null);
            this._textDeclaration.string = (content);
            G_Prompt.showTip(Lang.get('guild_declaration_change_success'));
        }
    }
    onButtonModifyName(sender) {
        if (!this._canModifyGuildName) {
            return
        }
        var callback = function (guildName) {
            G_UserData.getGuild().c2sGuildChangeName(guildName);
        };

        var popup: PopupGuildChangeName = Util.getNode('prefab/guild/PopupGuildChangeName', PopupGuildChangeName);
        popup.updateUI(Lang.get('guild_change_name_title'), Lang.get('guild_change_name_hint'), {
            type: TypeConvertHelper.TYPE_RESOURCE,
            value: DataConst.RES_DIAMOND,
            size: parseInt(G_ConfigLoader.getConfig(ConfigNameConst.PARAMETER).get(ParameterIDConst.GUILD_RENAME_COST)["content"])
        }, callback.bind(this));
        popup.openWithAction();
    }
    _dismissSuccess() {
        G_Prompt.showTip(Lang.get('guild_tip_dismiss_success'));
        G_SceneManager.popScene();
    }
    _updateList() {
        this._guildTaskList = G_UserData.getGuild().getMyGuild().getSortedTaskDataList();
        if (this._listItemSource2.content.childrenCount > 0) {
            for(let i = 0; i < this._listItemSource2.content.childrenCount;i++){
                if(this._guildTaskList[i]){
                    let cell = this._listItemSource2.content.children[i].getComponent(GuildTaskItemCell) as GuildTaskItemCell;
                    cell.updateUI(this._guildTaskList[i]);
                }
            }
        }
        else {
            this._listItemSource2.content.removeAllChildren();
            this._listItemSource2.content.height = 0;
            //更新任务的数据
            for (let i = 0; i < this._guildTaskList.length; i++) {
                if (this._guildTaskList[i]) {
                    var resource = cc.resources.get("prefab/guild/GuildTaskItemCell");
                    var node1 = cc.instantiate(resource) as cc.Node;
                    let cell = node1.getComponent(GuildTaskItemCell) as GuildTaskItemCell;
                    this._listItemSource2.content.addChild(cell.node);
                    cell.node.y = (i + 1) * -120;
                    cell.updateUI(this._guildTaskList[i]);
                    this._listItemSource2.content.setContentSize(this._listItemSource2.content.getContentSize().width, this._listItemSource2.content.getContentSize().height + 120);
                }
            }
            this._listItemSource2.scrollToTop();
        }
    }
    _onItemUpdate(item, index) {
        if (this._guildTaskList[index]) {
            item.updateItem(index, this._guildTaskList[index]);
        }
    }
    _onItemSelected(item, index) {
    }
    _onItemTouch(lineIndex, index) {
        var data = this._guildTaskList[index];
        if (!data) {
            return;
        }
    }
    _onEventGuildBoxReward(event, rewards) {
        if (!G_UserData.getGuild().isInGuild()) {
            return;
        }
        this._taskBoxHelper.refreshBoxView();
        if (rewards) {
            G_Prompt.showAwards(rewards);
        }
    }
    _onEventGuildBaseInfoUpdate(event) {
        if (!G_UserData.getGuild().isInGuild()) {
            return;
        }
        this._updateInfo();
    }
    _onEventGuildNameChange(event) {
        if (!G_UserData.getGuild().isInGuild()) {
            return;
        }
        G_Prompt.showTip(Lang.get('guild_tip_change_name_success'));
    }
    _onEventGuildGetUserGuild(event) {
        if (!G_UserData.getGuild().isInGuild()) {
            return;
        }
        this._updateInfo();
    }

}